src/aks-preview/azext_aks_preview/_podidentity.py

```diff
@@ -0,0 +1,151 @@
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import time
+
+from knack.log import get_logger
+from knack.util import CLIError
+
+from ._client_factory import get_auth_management_client
+from ._consts import (CONST_MANAGED_IDENTITY_OPERATOR_ROLE,
+                      CONST_MANAGED_IDENTITY_OPERATOR_ROLE_ID)
+from ._roleassignments import add_role_assignment
+
+logger = get_logger(__name__)
+
+
+def _is_pod_identity_addon_enabled(instance):
+    if not instance:
+        return False
+    if not instance.pod_identity_profile:
+        return False
+    return bool(instance.pod_identity_profile.enabled)
+
+
+def _ensure_pod_identity_addon_is_enabled(instance):
+    if not _is_pod_identity_addon_enabled(instance):
+        raise CLIError('The pod identity addon is not enabled for this managed cluster yet.\n'
+                       'To enable, run "az aks update --enable-pod-identity')
+
+
+def _ensure_pod_identity_kubenet_consent(network_profile, pod_identity_profile, customer_consent):
+    if not network_profile or not network_profile.network_plugin:
+        # invalid data
+        return
+    if network_profile.network_plugin.lower() != 'kubenet':
+        # not kubenet, no need to check
+        return
+
+    if customer_consent is None:
+        # no set this time, read from previous value
+        customer_consent = bool(
+            pod_identity_profile.allow_network_plugin_kubenet)
+
+    if not customer_consent:
+        raise CLIError(
+            "--enable-pod-identity-with-kubenet is required for "
+            "enabling pod identity addon when using Kubenet network plugin")
+    pod_identity_profile.allow_network_plugin_kubenet = True
+
+
+def _fill_defaults_for_pod_identity_exceptions(pod_identity_exceptions):
+    if not pod_identity_exceptions:
+        return
+
+    for exc in pod_identity_exceptions:
+        if exc.pod_labels is None:
+            # in previous version, we accidentally allowed user to specify empty pod labels,
+            # which will be converted to `None` in response. This behavior will break the extension
+            # when using 2021-10-01 version. As a workaround, we always back fill the empty dict value
+            # before sending to the server side.
+            exc.pod_labels = dict()
+
+
+def _fill_defaults_for_pod_identity_profile(pod_identity_profile):
+    if not pod_identity_profile:
+        return
+
+    _fill_defaults_for_pod_identity_exceptions(pod_identity_profile.user_assigned_identity_exceptions)
+
+
+def _update_addon_pod_identity(
+    instance,
+    enable,
+    pod_identities=None,
+    pod_identity_exceptions=None,
+    allow_kubenet_consent=None,
+    models=None,
+):
+    if not enable:
+        ManagedClusterPodIdentityProfile = models.get("ManagedClusterPodIdentityProfile")
```

Comment:
question: as the `models` is a dict, why use `get` here? But I still think it's better to define a typed struct (https://docs.python.org/3/library/dataclasses.html) here for better type checking.

Reply:
It's the same question as in: https://github.com/Azure/azure-cli-extensions/pull/4243#discussion_r772105207

Reply:
I have no preference for using `get` or `[]` to get the value, but using `get` can avoid throwing an exception when the value does not exist. Of course, adding an attribute to `None` after getting a non-existent object with `get` will also report an error.

So I think it should be covered by unit tests to ensure that no one is getting unexpected value from the dictionary.

As for `dataclasses`, the feature was introduced in python 3.7. Before that, it was a third-party library that needed to be installed explicitly. Since azure-cli still supports python 3.6, as a plug-in, it may be difficult for us to request cli-core to add this dependency.